File: app/src/main/java/ntktest/example/com/ntktest/data/RssSourceRepository.java
package ntktest.example.com.ntktest.data;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.util.List;

public class RssSourceRepository {
    private static RssSourceRepository instance;

    @NonNull
    private RssSourceDao sourceDao = RssDatabase.instance().sourceDao();

    @Nullable
    private List<RssSource> cache;

    @NonNull
    public static RssSourceRepository instance() {
        if (instance == null) {
            instance = new RssSourceRepository();
        }
        return instance;
    }

    @NonNull
    public List<RssSource> getSources(){
        return getSources(false);
    }

    public void add(@NonNull RssSource source) {
        sourceDao.insert(source);
        getSources(true);
    }

    public void update(@NonNull RssSource source) {
        sourceDao.update(source);
        getSources(true);
    }

    public void delete(@NonNull RssSource source) {
        sourceDao.delete(source);
        getSources(true);
    }

    private RssSourceRepository() {}

    private List<RssSource> getSources(boolean forceUpdate) {
        if (forceUpdate || cache == null) {
            cache = sourceDao.all();
        }
        return cache;
    }
}

File: app/src/main/java/com/example/rssreader/data/RssSourceRepository.kt
package com.example.rssreader.data

import com.example.rssreader.di.RssDatabaseFactory
import io.reactivex.Completable
import io.reactivex.Maybe
import io.reactivex.Single
import io.reactivex.schedulers.Schedulers

class RssSourceRepository(private val sourceDao: RssSourceDao) {
    fun getSources(): Single<List<RssSource>> =
        sourceDao.all()
            .subscribeOn(Schedulers.io())

    fun getById(id: Int): Maybe<RssSource> =
        sourceDao.getById(id)
            .subscribeOn(Schedulers.io())

    fun add(source: RssSource): Completable =
        sourceDao.insert(source)
            .subscribeOn(Schedulers.io())

    fun update(source: RssSource): Completable =
        sourceDao.update(source)
            .subscribeOn(Schedulers.io())

    fun delete(source: RssSource): Completable =
        sourceDao.delete(source)
            .subscribeOn(Schedulers.io())
}